onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc

```diff
@@ -0,0 +1,161 @@
+                             "Data type for starts and ends inputs' is not supported in this build. Got ",
+                             data_type);
+    }
+
+    return Status::OK();
+  };
+  ORT_RETURN_IF_ERROR(CopyInputData(1, input_starts));
+  ORT_RETURN_IF_ERROR(CopyInputData(2, input_ends));
+  ORT_RETURN_IF_ERROR(CopyInputData(3, input_axes));
+  ORT_RETURN_IF_ERROR(CopyInputData(4, input_steps));
+  ORT_RETURN_IF_ERROR(
+      SliceOp::PrepareForComputeHelper(input_starts, input_ends, input_axes, input_steps, compute_metadata));
+
+  std::transform(compute_metadata.starts_.cbegin(), compute_metadata.starts_.cend(),
+                 starts.begin(),
+                 [](int64_t i) { return SafeInt<uint32_t>(i); });
+  std::transform(compute_metadata.ends_.cbegin(), compute_metadata.ends_.cend(), compute_metadata.starts_.cbegin(),
+                 sizes.begin(),
+                 [](int64_t i, int64_t j) { return SafeInt<uint32_t>(i - j); });
+
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("slice", inputs,
+                                                                            emscripten::val::array(starts),
+                                                                            emscripten::val::array(sizes));
+
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(output));
+  return Status::OK();
+}
+
+bool SliceOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                                       const logging::Logger& logger) const {
+  const auto& name = node.Name();
+  const auto& op_type = node.OpType();
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    return false;
+  }
+  if (input_defs.size() == 5) { // Check steps.
+    const auto& steps_tensor = *initializers.at(input_defs[4]->Name());
+    std::vector<uint8_t> unpacked_tensor;
+    auto status = onnxruntime::utils::UnpackInitializerData(steps_tensor, unpacked_tensor);
+    if (!status.IsOK()) {
+      LOGS(logger, ERROR) << "Error while unpacking steps_tensor: " << status.ErrorMessage();
+      return false;
+    }
+    const auto data_type = steps_tensor.data_type();
+    // WebNN doesn't support steps other than 1.
+    if (data_type == ONNX_NAMESPACE::TensorProto_DataType_INT64) {
+      auto tensor_data = std::vector<int64_t>(reinterpret_cast<int64_t*>(unpacked_tensor.data()),
+                                              reinterpret_cast<int64_t*>(unpacked_tensor.data() +
+                                                                         unpacked_tensor.size()));
+      if (!std::all_of(tensor_data.begin(), tensor_data.end(), [](int64_t i) { return i == 1; })) {
+        return false;
+      }
+    } else if (data_type == ONNX_NAMESPACE::TensorProto_DataType_INT32) {
+      auto tensor_data = std::vector<int32_t>(reinterpret_cast<int32_t*>(unpacked_tensor.data()),
```

Comment:
Do we need to allocate a temporary vector if we're only using it to check `all_of`?

Does `if (!std::all_of(unpacked_tensor.data(), unpacked_tensor.data() + unpacked_tensor.size(), [](int32_t i) { return i == 1; }))` work?

Reply:
You could also use `any_of`, but both work, and I'm content either way:

if (std::any_of(tensor_data.begin(), tensor_data.end(), [](int32_t i) { return i != 1; }) {
}